ipvlan support for libnetwork

brb opened the pull request:
- [x] Fix `level=warning msg="Unable to setup ipvlan slave" containerID=d3209e6eaf error="failed to create clsact qdisc on \"cilium0\": file exists" maxRetry=20 retry=1 subsys=workload-watcher willRetry=false`.
- [ ] Fix restoration of endpoint after cilium-agent has restarted.
- [ ] Integration tests.
- [ ] Update guides.

EDIT: ^^ will be done in separate PRs.


<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6674)
<!-- Reviewable:end -->

brb committed:
endpoint/connector: Convert ipvlan mode in ipvlan setup routine

The same type conversion (from "string" to "netlink.IPvlanMode") would
be required for "plugin/cilium-docker". So, move it to the common
function.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
workloads: Add workload option for datapath mode

It's going to be used by Docker runtime workload event handler
to decide whether a BPF object needs to be attached to ipvlan
slave device.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
connector: Do not rename ipvlan iface if not target name is the same

In the case of libnetwork plugin, ipvlan iface name does not change,
so let's avoid calling netlink to save some CPU cycles.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
connector: Unify ipvlan iface name creation

Use the same function as for veth to generate deterministic
ipvlan iface name.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
endpoint: Add isDatapathMapPinned field

The field denotes whether the datapath map has been pinned for
the given endpoint.

In the case of libnetwork+ipvlan, this can be used to detect (hack)
whether all BPF-related initialization has been done for an ipvlan
slave device, as we will pin the map only afterwards.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

brb committed:
cilium-docker: Add support for ipvlan

This change enables ipvlan for the Cilium Docker libnetwork plugin.

In opposite to CNI, libnetwork moves container's netdev (ipvlan slave)
to container's network namespace by itself. This complicates our
implementation a lot, because the kernel resets qdisc's of the netdev
after the move. Therefore, the BPF object used for ipvlan gets removed.

To overcome this limitation, we split the ipvlan slave initialization
into two parts.

The first one creates the ipvlan slave in the host network namespace,
and it's executed in the JoinEndpoint context.

The second one attaches the BPF objects to the device in the container
network namespace, and is executed asynchronously by the Docker
workloads monitors.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

borkmann commented on pkg/endpoint/connector/add.go:
```diff
@@ -59,13 +59,6 @@ func randIfStr(num int) string {
-// given endpointID. This is similar to Endpoint2TempIfName() but uses a
-// random string instead of endpoint ID.
-func Endpoint2TempRandIfName() string {
-	return temporaryInterfacePrefix + "_" + randIfStr(5)
```
Nit: we should also get rid of randIfStr() as its only used there.

brb committed:
endpoint: rename method MapPin to MapPinLocked

This makes a user of the method aware that the lock should be taken
before calling the method.

Also, call the method in daemon/endpoint.go after acquiring the lock.

Signed-off-by: Martynas Pumputis <martynas@covalent.io>

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -2232,6 +2246,11 @@ func (e *Endpoint) MapPin() error {
 
 	err = bpf.ObjPin(mapFd, e.BPFIpvlanMapPath())
-	unix.Close(mapFd)
+	defer unix.Close(mapFd)
```
nit: The defer is closing a resource configured a few lines up, it's a bit easier to read if there's

```
doThing1()
defer releaseThing1()
if err = doThing2(); err != nil {
  return err
}
return nil
```

rather than

```
doThing1()
doThing2()
defer releaseThing1()
if err == nil {
  doSomethingElse()
}
return err
```

joestringer commented on pkg/workloads/docker.go:
```diff
@@ -487,6 +505,83 @@ func (d *dockerClient) handleCreateWorkload(id string, retry bool) {
+		// can be called many times for the same container in parallel), we can
+		// check whether the datapath map has been pinned which indicates previous
+		// successful invocation of the function for the same container.
+		if ep.GetDockerNetworkID() != "" && d.datapathMode == option.DatapathModeIpvlan {
```
Any chance this chunk of logic can be factored into a `configureIPVLAN()` function or similar? Looks like this function is currently a nice ~1 page long or so, and this seems to double the size of the function for handling one specific case.

joestringer commented on plugins/cilium-docker/driver/driver.go:
```diff
@@ -297,18 +299,38 @@ func (driver *driver) createEndpoint(w http.ResponseWriter, r *http.Request) {
+				}
+			}
+		}()
+	}
```
All of the error handling logic for both of these cases in the switch statement seems to be the same except for the device choice/name. I think it could be folded into one check after the switch statement, with the log messages including something like `device.Name` as a hint as to whether it was attempting veth or ipvlan mode.

joestringer reviewed (commented):
Looks fine, some minor code organisation suggestions below.
